doc/pins/PINS_SONiC_Design_for_SaiGenericExt.md

```diff
@@ -0,0 +1,429 @@
+  - For each retrieved field and parameter, check if any of them needs to be cross-referred to an entry in another table. Update reference count of dependent objects based on CRUD operation
+  - If field is a counter/meter directive, create respective SAI counter/meter object if one does not exist, or remove - SAI counter/meter object in case of delete event
+  - Prepare SAI Extension API parameters and call appropriate SAI Extension API
+
+
+#### Table processing precedence <a name="Table-processing-precedence"></a>
+SAI.P4 has well-defined ordering of functional blocks in the SAI pipeline. For the SAI pipeline relevant P4RT tables, P4Orch currently has a predefined order of processing APPL_DB entries belonging to those tables. This pre-defined table precedence helps avoid out-of-order processing of cross-dependent entries from a given batch. The same behavior of processing cross-dependent entries in a specific order of table precedence is necessary for entries across SAI Extension tables, and even across SAI path and SAI Extension path tables.
+
+For in-order processing of table entries across cross-dependent tables, a specific processing order of tables can be decided at the initialization time when the Table Definition database is built. After preparing a cross-reference state across all actions from all tables, a dependency graph can be built that can decide the processing order of tables. Note that this dependency graph is built once at the initialization time. At the time of draining table entries that are received from APPL_DB, P4Orch drains entries from tables in their precedence determined by pre-built dependency graph.
+
+The SAI Extension manager in the P4Orch, which caters to serving extension tables, maintains table entries separately per table. Each of these tables are tagged with an algorithmically determined precedence value based on dependency graph. At the time of draining table entries, The SAI Extension manager may be called multiple times, each time with a different precedence value and so allowing the SAI Extension manager process table entries only from those tables matching the precedence value.
+
+#### Cross Table dependencies <a name="Cross-Table-dependencies"></a>
+Once the Table Definition database is built, the state created for each table would have cross-reference table/match-fields mapped from eg. an action parameter if that action parameter is defined to have a cross table dependency. As stated earlier, this document focuses on dependencies from PINS SAI Extension table entries to another PINS SAI Extension table entry, and from PINS SAI Extension table entris to PINS SAI table entry. Thus, only those scenarios are considered in the design [This framework can be leveraged, with some enhancements, for other scenarios as well like for PINS object dependencies on non-PINS SAI objects, PINS SAI table entries referencing to entry from PINS SAI Extension table].

```

Comment:
It seems to me that this is actually the key weakness of the SAI extension manager design. SAI DB already has all the objects so building cross table design would have been straightforward, and as a day-1 scenario. However, with the SAI extension manager design, the objects basically need to have a copy in the SAI extension manager, which is missing for objects from other parts of the system at this point

Reply:
Hi Lihua, Since Sai API calls require OID to be passed as well, all the Orchs within SWSS store the OID mappings. Example being mirrororch, routeorch, vrforch. Similarly P4orch also maintains p4 entries/key to oid mapping as added last year and does ref counting as well. 
The new SAI-Extensions manager within P4Orch is extending the same functionality for handling the non-SAI objects and have the corresponding OID mappings for dependencies. Since the SAI objects are “known” it is possible for us to query the corresponding manager in OrchAgent to retrieve the OID info, but for SAI Generic Extensions to handle the same for new non-SAI objects we needed to add a generic way to retrieve OID given a key-value and here the dependency graph helps. Please let us know and we can add more details.
Thank you.
Reshma